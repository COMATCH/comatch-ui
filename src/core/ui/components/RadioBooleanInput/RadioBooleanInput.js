import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import { castStringToBoolean } from '../../../utils';
import { RadioGroupInput } from '../RadioGroupInput';

const propTypes = {
    disabled: PropTypes.bool,
    display: PropTypes.oneOf(['block', 'inline']),
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
    falseLabel: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    required: PropTypes.bool,
    value: PropTypes.bool,
    trueLabel: PropTypes.string.isRequired,
};

const defaultProps = {
    disabled: false,
    display: 'block',
    id: null,
    inputError: '',
    label: '',
    onChange: noop,
    required: false,
    value: null,
};

export const RadioBooleanInput = ({
    disabled,
    display,
    id,
    inputError,
    label,
    name,
    falseLabel,
    onChange,
    required,
    value,
    trueLabel,
}) => {
    const options = [{ label: trueLabel, value: true }, { label: falseLabel, value: false }];

    function handleChange(event) {
        onChange({ ...event, target: { ...event.target, value: castStringToBoolean(event.target.value) } });
    }

    return (
        <RadioGroupInput
            disabled={disabled}
            display={display}
            id={id}
            inputError={inputError}
            label={label}
            name={name}
            onChange={handleChange}
            options={options}
            required={required}
            value={value}
        />
    );
};

RadioBooleanInput.defaultProps = defaultProps;

RadioBooleanInput.propTypes = propTypes;
