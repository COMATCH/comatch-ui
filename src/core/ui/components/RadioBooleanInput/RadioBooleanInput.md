
## RadioBooleanInput

From [`RadioBooleanInput.js`](RadioBooleanInput.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**disabled** | `Boolean` | `false` | :x: | 
**display** | `Enum('block','inline')` | `'block'` | :x: | 
**falseLabel** | `String` |  | :white_check_mark: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `String` | `''` | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onChange** | `Function` | `noop` | :x: | 
**required** | `Boolean` | `false` | :x: | 
**trueLabel** | `String` |  | :white_check_mark: | 
**value** | `Boolean` | `null` | :x: | 



