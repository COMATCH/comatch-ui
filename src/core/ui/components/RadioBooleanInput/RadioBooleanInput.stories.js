import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { RadioBooleanInput } from './RadioBooleanInput';

class RadioBooleanInputContainer extends React.Component {
    state = { available: true };

    handleChange = (event) => {
        this.setState({ available: JSON.parse(event.target.value) });
        action('changed in container')(event.target.name, event.target.value);
    };

    render() {
        return (
            <RadioBooleanInput
                name="available"
                onChange={this.handleChange}
                value={this.state.available}
                trueLabel="yes"
                falseLabel="no"
            />
        );
    }
}

storiesOf('RadioBooleanInput', module)
    .add('with name', () => <RadioBooleanInput trueLabel="yes" falseLabel="no" name="choice" />)
    .add('with name and display inline', () => (
        <RadioBooleanInput name="choice" trueLabel="yes" falseLabel="no" display="inline" />
    ))
    .add('with name and display block', () => (
        <RadioBooleanInput name="choice" trueLabel="yes" falseLabel="no" display="block" />
    ))
    .add('with name and label', () => (
        <RadioBooleanInput name="choice" trueLabel="yes" falseLabel="no" label="I am the ultimate label" />
    ))
    .add('with name and error', () => (
        <RadioBooleanInput name="choice" trueLabel="yes" falseLabel="no" inputError="I am the ultimate error" />
    ))
    .add('with name, label and error', () => (
        <RadioBooleanInput
            name="choice"
            label="I am the ultimate label"
            trueLabel="yes"
            falseLabel="no"
            inputError="I am the ultimate error"
        />
    ))
    .add('with name, label and display inline', () => (
        <RadioBooleanInput
            name="choice"
            display="inline"
            trueLabel="yes"
            falseLabel="no"
            label="I am the ultimate label"
        />
    ))
    .add('with name, error and display inline', () => (
        <RadioBooleanInput
            name="choice"
            display="inline"
            trueLabel="yes"
            falseLabel="no"
            inputError="I am the ultimate error"
        />
    ))
    .add('with name, label, error and display inline', () => (
        <RadioBooleanInput
            name="choice"
            display="inline"
            label="I am the ultimate label"
            inputError="I am the ultimate error"
            trueLabel="yes"
            falseLabel="no"
        />
    ))
    .add('with onChange', () => {
        const handleChange = (event) => action('changed')(event.target.name, event.target.value);
        return <RadioBooleanInput name="choice" trueLabel="yes" falseLabel="no" onChange={handleChange} />;
    })
    .add('in a container', () => <RadioBooleanInputContainer />);
