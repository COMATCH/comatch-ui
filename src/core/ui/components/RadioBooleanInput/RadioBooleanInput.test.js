import React from 'react';
import renderer from 'react-test-renderer';
import noop from 'lodash/noop';
import { mount } from 'enzyme';
import { lorem } from 'faker';
import { RadioBooleanInput } from './RadioBooleanInput';
import '../../../config/tests/setup';

describe('RadioBooleanInput tests', () => {
    const REQUIRED_PROPS = {
        name: 'name',
        falseLabel: 'no',
        trueLabel: 'yes',
    };

    const OPTIONAL_PROPS = {
        disabled: false,
        display: 'block',
        id: null,
        inputError: 'InputError',
        label: lorem.words(),
        onChange: noop,
        required: false,
        value: null,
    };

    it('should render and check snapshot', () => {
        const radioBooleanInput = renderer.create(<RadioBooleanInput {...REQUIRED_PROPS} />).toJSON();
        expect(radioBooleanInput).toMatchSnapshot();
    });

    it('should render with only required props', () => {
        const props = {
            ...REQUIRED_PROPS,
        };
        const radioBooleanInput = mount(<RadioBooleanInput {...props} />);
        // Required props:
        expect(radioBooleanInput.prop('name')).toEqual(props.name);
        expect(radioBooleanInput.prop('falseLabel')).toEqual(props.falseLabel);
        expect(radioBooleanInput.prop('trueLabel')).toEqual(props.trueLabel);
        // Defaults:
        expect(radioBooleanInput.prop('disabled')).toEqual(false);
        expect(radioBooleanInput.prop('display')).toEqual('block');
        expect(radioBooleanInput.prop('id')).toEqual(null);
        expect(radioBooleanInput.prop('inputError')).toEqual('');
        expect(radioBooleanInput.prop('label')).toEqual('');
        expect(radioBooleanInput.prop('onChange')).toEqual(noop);
        expect(radioBooleanInput.prop('required')).toEqual(false);
        expect(radioBooleanInput.prop('value')).toEqual(null);
    });

    it('should render correctly with optional props', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };
        const radioBooleanInput = mount(<RadioBooleanInput {...props} />);
        // Required props:
        expect(radioBooleanInput.prop('name')).toEqual(props.name);
        expect(radioBooleanInput.prop('falseLabel')).toEqual(props.falseLabel);
        expect(radioBooleanInput.prop('trueLabel')).toEqual(props.trueLabel);
        // Defaults:
        expect(radioBooleanInput.prop('disabled')).toEqual(props.disabled);
        expect(radioBooleanInput.prop('display')).toEqual(props.display);
        expect(radioBooleanInput.prop('id')).toEqual(null);
        expect(radioBooleanInput.prop('inputError')).toEqual(props.inputError);
        expect(radioBooleanInput.prop('label')).toEqual(props.label);
        expect(radioBooleanInput.prop('onChange')).toEqual(noop);
        expect(radioBooleanInput.prop('required')).toEqual(props.required);
        expect(radioBooleanInput.prop('value')).toEqual(null);
    });

    describe('verifying the HTML structure', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };
        it('should render with `trueLabel` text message', () => {
            const radioBooleanInput = mount(<RadioBooleanInput {...props} />);
            expect(radioBooleanInput);
        });

        it('should expect nodes/dom elements', () => {
            const radioBooleanInput = mount(<RadioBooleanInput {...props} />);
            expect(radioBooleanInput).toHaveLength(1);
            expect(radioBooleanInput.find('div')).toHaveLength(5);
            expect(radioBooleanInput.find('input')).toHaveLength(2);
            expect(radioBooleanInput.find('input').at(0).prop('value')).toBe(true);
            expect(radioBooleanInput.find('input').at(1).prop('value')).toBe(false);

            expect(radioBooleanInput.find('label')).toHaveLength(3);
            expect(radioBooleanInput.find('label').at(0).text()).toEqual(props.label);
            expect(radioBooleanInput.find('label').at(1).text()).toEqual(props.trueLabel);
            expect(radioBooleanInput.find('label').at(2).text()).toEqual(props.falseLabel);

            expect(radioBooleanInput.find('span')).toHaveLength(3);
            expect(radioBooleanInput.find('span').at(0).text()).toEqual(props.trueLabel);
            expect(radioBooleanInput.find('span').at(1).text()).toEqual(props.falseLabel);
            expect(radioBooleanInput.find('span').at(2).text()).toEqual(props.inputError);
        });
    });
});
