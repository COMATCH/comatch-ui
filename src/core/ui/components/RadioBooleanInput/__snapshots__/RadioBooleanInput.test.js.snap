// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioBooleanInput tests should render and check snapshot 1`] = `
<div
  className="Input RadioGroupInput"
  id={null}
>
  
  <div
    className="RadioInput block"
    id={null}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        id={null}
        name="name"
        onChange={[Function]}
        type="radio"
        value={true}
      />
      <div
        className="RadioInput__input"
        role="radio"
      />
      <span
        className="InputLabel RadioInput__label sc-bdVaJa eCYGDq"
        id={null}
      >
        yes
      </span>
    </label>
    
  </div>
  <div
    className="RadioInput block"
    id={null}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        id={null}
        name="name"
        onChange={[Function]}
        type="radio"
        value={false}
      />
      <div
        className="RadioInput__input"
        role="radio"
      />
      <span
        className="InputLabel RadioInput__label sc-bdVaJa eCYGDq"
        id={null}
      >
        no
      </span>
    </label>
    
  </div>
  
</div>
`;
